docs/specs/account-linking.md

```diff
@@ -120,35 +120,62 @@ We define the action to link the new oauth identity with the existing identity's
 
 Currently, only `error` and `login_and_link` will be implemented.
 
-### Defining a default for all flows
+### Override the config in specific flow
+
+We provide a way to override account linking configs in a specific flow. This is because there are valid use cases that each flow should behave differently. For example, you may probably want the `default` signup flow to do account linking with the `default` signup flow, but in other side, you want another signup flow `signup_flow_1` to do account linking with another `login_flow_1` login_flow. Use the following config to achieve this behavior:
 
-All config mentioned above was defined inside a single flow object of a `signup_flows`. However, you may simply want a config applied to all signup flows. Therefore, we have the `default_account_linking` section for this purpose. Here is an example:
 
 ```yaml
 authentication_flow:
-  default_account_linking:
+  account_linking:
     oauth:
-      - alias: google
+      - alias: adfs
         oauth_claim:
-          pointer: "/email"
+          pointer: "/preferred_username"
         user_profile:
-          pointer: "/email"
+          pointer: "/preferred_username"
         action: login_and_link
         login_flow: default
+  signup_flows:
+    - name: signup_flow_1
+      account_linking:
+        oauth:
+          - alias: adfs # This will override the account_linking configs for adfs above 
+            action: link_without_login_when_verified
+            login_flow: login_flow_1
+      steps:
+        - name: identify
+          type: identify
+          one_of:
+            - identification: email
+              steps:
+                - name: authenticate_primary_email
+                  type: create_authenticator
+                  one_of:
+                    - authentication: primary_password
+            - identification: oauth
```

Comment:
Actually this design is a bit uncomfortable to me.
Considering this example:

```yaml
authentication_flow:
  account_linking:
    oauth:
      - alias: adfs
        oauth_claim:
          pointer: "/preferred_username"
        user_profile:
          pointer: "/preferred_username"
        action: login_and_link
        login_flow: default
      - alias: adfs
        oauth_claim:
          pointer: "/email"
        user_profile:
          pointer: "/email"
        action: login_and_link
        login_flow: default
  signup_flows:
    - name: signup_flow_1
      account_linking:
        oauth:
          - alias: adfs # This will override the account_linking configs for adfs above 
            action: link_without_login_when_verified
            login_flow: login_flow_1
      steps:
        - name: identify
          type: identify
          one_of:
            - identification: email
              steps:
                - name: authenticate_primary_email
                  type: create_authenticator
                  one_of:
                    - authentication: primary_password
            - identification: oauth
```

The problem here is:
1. One object inside `signup_flows.account_linking.oauth` overridden two object inside `authentication_flow.account_linking`. Because the two object are both `alias: adfs`.
2. Cannot define different `action` for different pointers.

For 1, I am thinking about if we should limit the number of config per alias to 1 only, but this seems making an assumption that one provider should only have one logic (One pair of pointers) to link accounts.

For 2, it seems acceptable as there should be no normal usecase to do different action when using different pointer?

Reply:
Another option would be, just keep `oauth_claim` and `user_profile` in the flow specific config. And each if you want to override something, you should copy the whole object from `authentication_flow.account_linking`.

Reply:
Given

1. We expect the matching logic is consistent for the whole project. Thus the following part should be shared by the whole project

```yaml
# No login_flow in the item. So the default is to use the login_flow of the same name.
account_linking:
  oauth:
  - alias: adfs
    oauth_claim:
      pointer: "/preferred_username"
    user_profile:
      pointer: "/preferred_username"
    action: login_and_link
  - alias: adfs
    oauth_claim:
      pointer: "/email"
    user_profile:
      pointer: "/email"
    action: login_and_link
```

2. For each match, we want to allow the specific signup flow / promote flow to specify the action, and the login flow to use. Thus, I think the items are missing a name attribute for other part of the config to reference them. Once we have name, I think we should flatten the items into a single list.

```yaml
# This is now moved to top-level, not under authentication_flow
account_linking:
- name: asdf_preferred_username
  type: oauth
  alias: adfs
  oauth_claim:
    pointer: "/preferred_username"
  user_profile:
    pointer: "/preferred_username"
  action: login_and_link
- name: asdf_email
  type: oauth
  alias: adfs
  oauth_claim:
    pointer: "/email"
  user_profile:
    pointer: "/email"
  action: login_and_link
- name: email
  type: login_id
  key: email
  action: login_and_link
```

3. In case the signup flow wants to override, use the name to reference the item

```yaml
signup_flows:
- name: default
  account_linking:
    override:
    - name: asdf_preferred_username
       login_flow: another_login_flow
       action: link_without_login_when_verified
```